src/assets-generator.ts

```diff
@@ -0,0 +1,355 @@
+        ].join("\n"),
+      );
+      return;
+    }
+  }
+
+  const useImage = Array.isArray(images) ? images[0] : images;
+  const imageFile = resolve(root, useImage);
+  const publicDir = resolve(root, ctx.viteConfig.publicDir ?? "public");
+  const outDir = ctx.clientOutBaseDir;
+  const imageName = relative(publicDir, imageFile);
+  const imageOutDir = dirname(resolve(outDir, imageName));
+
+  const xhtml = userHeadLinkOptions?.xhtml === true;
+  const includeId = userHeadLinkOptions?.includeId === true;
+  const assetsInstructions = await instructions({
+    imageResolver: () => readFile(resolve(root, useImage)),
+    imageName,
+    preset,
+    faviconPreset: userHeadLinkOptions?.preset,
+    htmlLinks: { xhtml, includeId },
+    basePath: ctx.basePathRelDir || "/",
+    resolveSvgName:
+      userHeadLinkOptions?.resolveSvgName ?? ((name) => basename(name)),
+  });
+  const {
+    includeWebManifest = true,
+    includeHtmlHeadLinks = true,
+    includeThemeColor = false,
+    overrideManifestIcons = false,
+  } = ctx.options.assets!;
+
+  if (assetContext === undefined) {
+    return {
+      lastModified: Date.now(),
+      assetsInstructions,
+      cache: new Map<string, ResolvedIconAsset>(),
+      useImage,
+      imageFile,
+      publicDir,
+      outDir,
+      imageName,
+      imageOutDir,
+      xhtml,
+      includeId,
+      // normalize sources
+      sources: sources.map((source) => source.replace(/\\/g, "/")),
+      includeWebManifest,
+      includeThemeColor,
+      includeHtmlHeadLinks,
+      overrideManifestIcons,
+    } satisfies AssetsGeneratorContext;
+  }
+
+  assetContext.lastModified = Date.now();
+  assetContext.assetsInstructions = assetsInstructions;
+  assetContext.useImage = useImage;
+  assetContext.imageFile = imageFile;
+  assetContext.outDir = outDir;
+  assetContext.imageName = imageName;
+  assetContext.imageOutDir = imageOutDir;
+  assetContext.xhtml = xhtml;
+  assetContext.includeId = includeId;
+  assetContext.includeWebManifest = includeWebManifest;
+  assetContext.includeThemeColor = includeThemeColor;
+  assetContext.includeHtmlHeadLinks = includeHtmlHeadLinks;
+  assetContext.overrideManifestIcons = overrideManifestIcons;
+  assetContext.cache.clear();
+}
+
+async function readManifestFile({ options, viteConfig }: QwikPWAContext) {
+  if (
+    !options.assets!.overrideManifestIcons ||
+    !options.assets!.includeWebManifest
+  )
+    return;
+
+  const name =
+    typeof options.assets!.includeWebManifest === "string"
+      ? options.assets!.includeWebManifest
+      : "manifest.webmanifest";
+
+  const manifestFile = resolve(viteConfig.publicDir ?? "public", name);
+  const isFile = await lstat(manifestFile)
+    .then((stat) => stat.isFile())
+    .catch(() => false);
+  if (!isFile) return;
+
+  return {
+    manifestFile,
+    manifest: await readFile(manifestFile, { encoding: "utf-8" }).then(
+      JSON.parse,
+    ),
+  };
+}
```

Comment:
Is there any way we avoid this complexity? any library maybe? So we remove this file or make it much much more smaller.

Reply:
I don't have a library for it, the logic for vite is avalaible only in vite-plugin-pwa, other meta frameworks should provide its own wrapper